src/content/docs/en/guides/integrations-guide/react.mdx

```diff
@@ -114,6 +114,66 @@ To use your first React component in Astro, head to our [UI framework documentat
 * 💧 client-side hydration options, and
 * 🤝 opportunities to mix and nest frameworks together
 
+## Integrate Actions with `useActionState()` (experimental)
+
+The `@astrojs/react` integration provides two experimental functions for use with [Astro Actions][astro-actions]: `experimental_withState()` and `experimental_getActionState()`.
+
+These are used with [React's useActionState() hook](https://react.dev/reference/react/useActionState) to read and update client-side state when triggering actions during form submission.
+
+To do so, pass `experimental_withState()` receiving the action you want to trigger in the `useActionState()` hook:
+
+```jsx title="Like.tsx" ins={2,6}
+import { actions } from 'astro:actions';
+import { experimental_withState } from '@astrojs/react/actions';
+
+export function Like({ postId }: { postId: string }) {
+  const [state, action, pending] = useActionState(
+    experimental_withState(actions.like),
+    0, // initial likes
+  );
+
+  return (
+    <form action={action}>
+      <input type="hidden" name="postId" value={postId} />
+      <button disabled={pending}>{state} ❤️</button>
+    </form>
+  );
+}
+```
+
+The `experimental_withState()` function will match the action's types with React's expectations and preserve metadata used for progressive enhancement, allowing it to work even when JavaScript is disabled on the user's device.
+
+### `experimental_getActionState()`
+
+<p>
+
+**Type:** `(context: APIContext) => Promise<T><br />
+<Since v="4.3.2" />
+</p>
+
+You can access the state stored by `useActionState()` on the server in your action `handler` with `experimental_getActionState()`. It accepts the Astro API context, and optionally, you can apply a type to the result. 
+
+The example below gets the current value of likes from a counter, typed as a number, in order to create an incrementing `like` action:
+
+```ts title="actions.ts" ins={3,11}
+import { defineAction } from 'astro:actions';
+import { z } from 'astro:schema';
+import { experimental_getActionState } from '@astrojs/react/actions';
+
+export const server = {
+  like: defineAction({
+    input: z.object({
+      postId: z.string(),
+    }),
+    handler: async ({ postId }, ctx) => {
+      const currentLikes = await experimental_getActionState<number>(ctx);
+      // write to database
+      return currentLikes + 1;
+    },
+  })
+};
+```
```

Comment:
````suggestion
```ts title="actions.ts" ins={3,11}
import { defineAction, type SafeResult } from 'astro:actions';
import { z } from 'astro:schema';
import { experimental_getActionState } from '@astrojs/react/actions';

export const server = {
  like: defineAction({
    input: z.object({
      postId: z.string(),
    }),
    handler: async ({ postId }, ctx) => {
      const currentLikes = await experimental_getActionState<SafeResult<any, number>>(ctx);
      // write to database
      return currentLikes + 1;
    },
  })
};
```
````